azkaban-exec-server/src/main/java/azkaban/execapp/FlowPreparer.java

```diff
@@ -0,0 +1,180 @@
+  private final Map<Pair<Integer, Integer>, ProjectVersion> installedProjects;
+  private final ProjectLoader projectLoader;
+
+  public FlowPreparer(ProjectLoader projectLoader,
+      File executionsDir,
+      File projectsDir,
+      Map<Pair<Integer, Integer>, ProjectVersion> installedProjects) {
+    this.projectLoader = projectLoader;
+    this.executionsDir = executionsDir;
+    this.projectsDir = projectsDir;
+    this.installedProjects = installedProjects;
+  }
+
+  /**
+   * Prepare the flow directory for execution.
+   *
+   * @param flow Executable Flow instance.
+   */
+  void setup(ExecutableFlow flow) {
+    File execDir = null;
+    try {
+      // First get the ProjectVersion
+      final ProjectVersion projectVersion = getProjectVersion(flow);
+
+      // Setup the project
+      setupProject(projectVersion);
+
+      // Create the execution directory
+      execDir = createExecDir(flow);
+
+      // Create the symlinks from the project
+      copyCreateHardlinkDirectory(projectVersion.getInstalledDir(), execDir);
+
+      log.info(String.format("Flow Preparation complete. [execid: %d, path: %s]",
+          flow.getExecutionId(), execDir.getPath()));
+    } catch (Exception e) {
+      log.error("Error in setting up project directory: " + projectsDir + ", Exception: " + e);
+      cleanup(execDir);
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Prepare the project directory.
+   *
+   * @param pv ProjectVersion object
+   * @throws ProjectManagerException
+   * @throws IOException
+   */
+  @VisibleForTesting
+  void setupProject(final ProjectVersion pv)
+      throws ProjectManagerException, IOException {
+    final int projectId = pv.getProjectId();
+    final int version = pv.getVersion();
+
+    final String projectDir = String.valueOf(projectId) + "." + String.valueOf(version);
+    if (pv.getInstalledDir() == null) {
+      pv.setInstalledDir(new File(projectsDir, projectDir));
+    }
+
+    // If directory exists. Assume its prepared and skip.
+    if (pv.getInstalledDir().exists()) {
+      log.info("Project already cached. Skipping download. " + pv);
+      return;
+    }
+
+    log.info("Preparing Project: " + pv);
+
+    File tempDir = new File(projectsDir, "_temp." + projectDir + "." + System.currentTimeMillis());
+
+    // TODO Why mkdirs? This path should be already set up.
+    tempDir.mkdirs();
+
+    ProjectFileHandler projectFileHandler = null;
+    try {
+      projectFileHandler = requireNonNull(projectLoader.getUploadedFile(projectId, version));
+      checkState("zip".equals(projectFileHandler.getFileType()));
+
+      log.info("Downloading zip file.");
+      final File zipFile = requireNonNull(projectFileHandler.getLocalFile());
+      final ZipFile zip = new ZipFile(zipFile);
+      Utils.unzip(zip, tempDir);
+
+      Files.move(tempDir.toPath(), pv.getInstalledDir().toPath(), StandardCopyOption.ATOMIC_MOVE);
+
+      log.warn(String.format("Project Preparation complete. [%s]", pv));
```

Comment:
Got it. Warning seems to me like a situation that may not be normal or dangerous.  In this case, this is routine and normal. I don't have a strong opinion either way. We can adjust after we gain more experience working with the logs.